framework/keyring-interface.md

```diff
@@ -75,66 +86,65 @@ perform the [Encrypt Data Key](#encrypt-data-key) behavior.
 
-Encrypt Data Key MAY modify the following fields in the [encryption materials](#structures.md#encryption-materials):
+Encrypt Data Key MAY populate the following fields in the [data key materials](#structures.md#data-key-materials):
 
 - [encrypted data keys](#structures.md#encrypted-data-keys)
 - [keyring trace](#structures.md#keyring-trace)
 
 To perform this behavior, the keyring creates one or more [encrypted data keys](#structures.md#encrypted-data-key)
-using the plaintext data key from the [encryption materials](#structures.md#encryption-materials) as input,
-and appends the [encrypted data keys](#structures.md#encrypted-data-key) to the encrypted data key list
-in the [encryption materials](#structures.md#encryption-materials).
+using the plaintext data key as input,
+and includes the [encrypted data keys](#structures.md#encrypted-data-key)
+in the [data key materials](#structures.md#data-key-materials) result.
 
 The [encrypted data keys](#structures.md#encrypted-data-key) produced by this keyring MUST
 have [ciphertexts](#structures.md#ciphertext) that can be decrypted to the plaintext data key in the
-[encryption materials](#structures.md#encryption-materials).
+[data key materials](#structures.md#data-key-materials).
 
-If OnEncrypt updates the [encryption materials](#data-structure.md#encryption-materials) with at least
-one new [encrypted data key](#structures.md#encrypted-data-key),
+If OnEncrypt outputs [data key materials](#structures.md#data-key-materials) with at least
+one [encrypted data key](#structures.md#encrypted-data-key),
 the [keyring trace](#structures.md#keyring-trace) returned by OnEncrypt MUST include at least one trace
 associated with this keyring that has the [ENCRYPTED DATA KEY](#structures.md#encrypted-data-key) flag.
 Note that this trace MAY include more than one flag,
 for example the [SIGNED ENCRYPTION CONTEXT flag](#structures.md#signed-encryption-context).
 
 ### OnDecrypt
 
-This interface takes [decryption materials](#structures.md#decryption-materials) and
-a list of [encrypted data keys](#structures.md#encrypted-data-key) as input and
-MAY modify it with the following behavior:
+The following inputs are REQUIRED:
+
+- [Algorithm Suite](#algorithm-suites.md)
+- [Encryption Context](#structures.md#encryption-context)
+- [Encrypted Data Keys](#structures.md#encrypted-data-keys)
+
+This interface MAY perform the following behavior:
 
 - [Decrypt data key](#decrypt-data-key)
 
-If this keyring attempted the above behavior, and succeeded, it MUST output the modified [decryption materials](#structures.md#decryption-materials).
+If this keyring attempted the above behavior, and succeeded, it MUST output the resulting [data key materials](#structures.md#data-key-materials). The [algorithm suite](#algorithm-suites.md) in the result must match the input [algorithm suite](#algorithm-suites.md), and the [encrypted data keys](#structures.md#encrypted-data-keys)
+list must be empty.
```

Comment:
> and the encrypted data keys list must be empty

What? This sounds like you are saying that a keyring would modify the EDK list. That would be a HUGE breaking change and I haven't seen any justification for it...